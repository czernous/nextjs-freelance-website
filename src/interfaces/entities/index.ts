export interface IResponsiveImage {
  width: number;
  url: string;
}

interface IObjectId {
  timestamp: number;
  machine: number;
  pid: number;
  increment: number;
  creationTime: Date | string;
}

interface IEntity {
  id: IObjectId;
  _id: string;
  createdAt: string;
  updatedAt: string;
}

interface IOpenGraph {
  title: string;
  description: string;
  imageUrl: string;
  url?: string;
  type?: string;
}

export interface ISeo {
  metaDescription: string;
  metaKeywords: string;
  openGraph: IOpenGraph;
}

interface ICategory extends IEntity {
  name: string;
}

interface IPost extends IEntity {
  title: string;
  categories: ICategory[];
  imageUrl: string;
  responsiveImgs: IResponsiveImage[];
  meta: ISeo; // seo
  isPublished: boolean;
  body: string;
}

export interface IImage extends IEntity {
  publicId: string;
  version: number;
  signature: string;
  width: number;
  height: number;
  format: string;
  resourceType: string;
  bytes: number;
  type: string;
  url: string;
  secureUrl: string;
  usedInPost?: IPost;
  responsiveUrls: IResponsiveImage[];
  thumbnailUrl: string;
  blurredImageUrl: string;
  path: string;
  altText?: string;
}

// FE only entities

interface IPage {
  updatedAt?: string;
}

export interface IHomePage extends IPage {
  ctaHeadline: string;
  ctaSubheadline: string;
  ctaBtnText: string;
  ctaBtnHref: string;
  imageUrl?: string;
  responsiveImgs?: IResponsiveImage[];
  meta: ISeo;
}

export interface IServicesPage extends IPage {
  content: string;
  meta: ISeo;
}

export interface IAboutPage extends IPage {
  description: string;
  imageUrl: string;
  responsiveImgs?: IResponsiveImage[];
  meta: ISeo;
}

export interface IContactPage extends IPage {
  description?: string;
  meta: ISeo;
}
