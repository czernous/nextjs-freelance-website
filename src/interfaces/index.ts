import { Color, Size } from '../enums';

export interface IButtonProps {
  buttonColor: Color;
  buttonSize: Size;
  buttonType: 'button' | 'submit' | 'reset';
  buttonStyle: 'primary' | 'secondary' | 'main-cta';
  buttonText: string;
  buttonHref?: string;
  buttonTarget?: string;
  buttonFullWidth: boolean;
  hasShadow: boolean;
}
