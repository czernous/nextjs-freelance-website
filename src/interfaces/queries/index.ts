export type IHttpMethod = 'GET' | 'POST' | 'PUT' | 'DELETE' | 'PATCH';

interface INextApiServerResponseError {
  error: {
    message: string | null;
  };
}

export interface INextApiServerResponse {
  ok: boolean;
  data: unknown | null;
  code: number;
  error: INextApiServerResponseError;
}
