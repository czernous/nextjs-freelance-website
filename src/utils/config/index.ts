import getConfig from 'next/config';

export function getBackendUrl() {
  const {
    serverRuntimeConfig: { BACKEND_URL },
  } = getConfig();
  return BACKEND_URL;
}

export function getBackendApiKey() {
  const {
    serverRuntimeConfig: { API_KEY },
  } = getConfig();
  return API_KEY;
}
