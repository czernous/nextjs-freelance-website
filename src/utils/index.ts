export * from './data-fetching';
export * from './testing';
