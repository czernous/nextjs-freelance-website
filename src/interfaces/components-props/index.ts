import { Color, InputTypes, Size } from '../../enums';
import { IImage, ISeo } from '../entities';

export interface IButtonProps {
  buttonColor: Color;
  buttonSize: Size;
  buttonType: 'button' | 'submit' | 'reset';
  buttonStyle: 'primary' | 'secondary' | 'main-cta';
  buttonText: string;
  buttonHref?: string;
  buttonTarget?: string;
  buttonFullWidth: boolean;
  hasShadow: boolean;
  isDisabled?: boolean;
  customClass?: string;
  onClick?: () => void;
}

export interface IInputProps {
  type: 'input' | 'textarea';
  inputType?: InputTypes;
  inputAdditionalClasses?: string;
  inputColor: Color;
  inputId: string;
  inputPlaceholder: string;
  isRequired?: boolean;
  isDisabled?: boolean;
  ariaDescribedBy?: string;
}

export interface IArticleCard {
  image: IImage;
  title: string;
  description: string;
  ctaText: string;
  ctaUrl: string;
  unoptimized: boolean;
}

export interface IPaginationButtonProps {
  text: string;
  action: string;
  active: boolean;
  disabled: boolean;
}

export interface INavbarProps {
  companyName: string;
  drawerWidth: number;
  navItems: INavItem[];
}

export interface INavItem {
  uuid: string;
  text: string;
  url: string;
}

export interface IClientPageLayoutProps {
  pageTitle: string;
  appTitle: string;
  meta: ISeo;
}

export interface ICustomSnackbarProps {
  text: string | null;
  severity: 'success' | 'error' | 'warning' | null;
  clearPropsFn?: ({ ...props }: ICustomSnackbarProps | null) => void;
}
