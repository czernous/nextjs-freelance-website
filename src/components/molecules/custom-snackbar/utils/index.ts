import { ICustomSnackbarProps } from '@src/interfaces/components-props';
import { SetStateAction } from 'react';

export const updateSnackbarProps = async (
  r: Response,
  callback: (value: SetStateAction<ICustomSnackbarProps | null>) => void,
) => {
  if (!r) return;
  const severity =
    r.status === 200 || r.status === 204 || r.status === 201
      ? 'success'
      : 'error';
  const hasJson = r.headers.get('Content-Type')?.includes('application/json');
  /* istanbul ignore next */
  const statusMessageText = async (r: Response) => {
    switch (r.status) {
      case 204:
        return 'Successfully saved data';
      case 201:
        return 'Successfully created';
      default:
        return await r.text();
    }
  };

  let statusMsg;

  try {
    statusMsg =
      hasJson && r.headers.get('Method') === 'PUT'
        ? await r.json()
        : {
            message: await statusMessageText(r),
          };
  } catch (error) {
    /* istanbul ignore next */
    statusMsg = hasJson
      ? {
          message:
            statusMsg.message ||
            (r.status === 200
              ? 'Successfully saved data'
              : 'Error saving data'),
        }
      : { message: 'Error reading response as text' };
  }

  callback({
    severity,
    text: statusMsg?.message ?? statusMsg,
  });
};
