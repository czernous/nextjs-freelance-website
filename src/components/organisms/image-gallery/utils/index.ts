import { IImage } from '@src/interfaces';
import { Dispatch, SetStateAction } from 'react';

/* istanbul ignore next */
export const getImages = async (): Promise<IImage[]> => {
  const imageResponse = await fetch(
    `${new URL(
      '/api/blog-data',
      window.location.origin,
    )}?url=/images&method=GET`,
  );

  const r = await imageResponse?.json();

  return r?.data;
};

/* istanbul ignore next */
export const handleGalleryOpen = async (
  setImages: Dispatch<SetStateAction<IImage[] | null>>,
) => {
  const imageResponse = await fetch(
    `${new URL(
      '/api/blog-data',
      window.location.origin,
    )}?url=/images&method=GET`,
  );

  const imagesData = await imageResponse?.json();

  setImages(imagesData?.data);
};

/* istanbul ignore next */
export const handleImageDelete = async (imageId: string) => {
  const imageDeleteResponse = await fetch(
    `${new URL(
      '/api/blog-data',
      window.location.origin,
    )}?url=/images/${imageId}&method=DELETE`,
  );

  if (imageDeleteResponse?.status === 204) {
    // TODO: add toast notification upon delete
  }

  if (imageDeleteResponse?.status === 404) {
    alert(
      'This image does not exist, you have probably deleted it already. Please reopen the gallery',
    );

    // TODO: add toast notification upon delete
  }
};
