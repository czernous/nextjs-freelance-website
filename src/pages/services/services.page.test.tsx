import { render, RenderResult, screen } from '@testing-library/react';
import { axe, toHaveNoViolations } from 'jest-axe';
import '@testing-library/jest-dom';
import Services from './index.page';
import { act } from 'react-dom/test-utils';
import React from 'react';

import { IServicesPage } from '../../interfaces';
import { mockNextRouter } from '../../utils';

expect.extend(toHaveNoViolations);

jest.mock('next/config', () => () => ({
  serverRuntimeConfig: {
    PROJECT_ROOT: __dirname,
  },
}));

const data: IServicesPage = {
  pageFields: {
    content:
      '<h2>Service #1</h2><p>Lorem ipsum dolor sit amet, consectetur adipiscing elit. Curabitur quis nunc libero. Nullam tincidunt lectus ut maximus dignissim. Mauris et risus in metus ornare convallis. Nullam in orci leo. Aenean sit amet arcu vitae ipsum mollis blandit. Pellentesque pellentesque sapien mollis, pharetra magna sit amet, varius eros. Donec eu dui ullamcorper ligula rutrum tincidunt. Sed eu venenatis mi. Nullam rutrum sagittis tortor, id feugiat dolor commodo a. Nullam venenatis rutrum elementum. Vestibulum vehicula lectus in nisl porta scelerisque at id metus. Sed porttitor purus eu cursus porta. Donec suscipit semper ligula, ut eleifend nunc commodo ut. Aenean sed mauris molestie, gravida nisi et, lobortis sapien. Nullam et massa mi. </p> <p>Nulla a sem sit amet nisl interdum tincidunt. Fusce pellentesque nec risus et lacinia. Donec dignissim enim ac vehicula malesuada. Lorem ipsum dolor sit amet, consectetur adipiscing elit. Etiam at varius augue. Phasellus varius, sem ac posuere mattis, augue turpis sollicitudin arcu, ut tincidunt arcu tellus vel sapien. Nulla sagittis auctor velit, ut tristique nulla sollicitudin et. Duis ac purus efficitur, congue purus egestas, ultricies quam. Nulla tincidunt nibh risus, vel semper lacus consectetur quis. Aenean luctus enim in justo rhoncus dignissim. Donec pellentesque fermentum sapien vitae aliquam.</p><h2>Service #1</h2><p>Lorem ipsum dolor sit amet, consectetur adipiscing elit. Curabitur quis nunc libero. Nullam tincidunt lectus ut maximus dignissim. Mauris et risus in metus ornare convallis. Nullam in orci leo. Aenean sit amet arcu vitae ipsum mollis blandit. Pellentesque pellentesque sapien mollis, pharetra magna sit amet, varius eros. Donec eu dui ullamcorper ligula rutrum tincidunt. Sed eu venenatis mi. Nullam rutrum sagittis tortor, id feugiat dolor commodo a. Nullam venenatis rutrum elementum. Vestibulum vehicula lectus in nisl porta scelerisque at id metus. Sed porttitor purus eu cursus porta. Donec suscipit semper ligula, ut eleifend nunc commodo ut. Aenean sed mauris molestie, gravida nisi et, lobortis sapien. Nullam et massa mi. </p> <p>Nulla a sem sit amet nisl interdum tincidunt. Fusce pellentesque nec risus et lacinia. Donec dignissim enim ac vehicula malesuada. Lorem ipsum dolor sit amet, consectetur adipiscing elit. Etiam at varius augue. Phasellus varius, sem ac posuere mattis, augue turpis sollicitudin arcu, ut tincidunt arcu tellus vel sapien. Nulla sagittis auctor velit, ut tristique nulla sollicitudin et. Duis ac purus efficitur, congue purus egestas, ultricies quam. Nulla tincidunt nibh risus, vel semper lacus consectetur quis. Aenean luctus enim in justo rhoncus dignissim. Donec pellentesque fermentum sapien vitae aliquam.</p>',
  },
  slug: 'services',
  meta: {
    metaDescription: 'string',
    metaKeywords: 'string',
    openGraph: {
      title: 'string',
      description: 'string',
      imageUrl: 'string',
    },
  },

  updatedAt: '2023-02-16T07:05:35.214Z',
};
let component: RenderResult;

beforeEach(async () => {
  mockNextRouter();
  component = render(<Services data={data} />);
});

describe('Services', () => {
  it('renders a heading (h1 - title)', () => {
    const heading = screen.getAllByRole('heading');

    expect(heading[0]).toBeInTheDocument();
  });

  it('renders homepage unchanged', () => {
    const { container } = component;
    expect(container).toMatchSnapshot();
  });

  it('has no axe violations', async () => {
    const { container } = component;
    await act(async () => {
      expect(await axe(container)).toHaveNoViolations();
    });
  });
});
