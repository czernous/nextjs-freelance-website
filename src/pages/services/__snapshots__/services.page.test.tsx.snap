// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Services renders homepage unchanged 1`] = `
<div>
  <div
    class="rich-text"
  >
    <h2>
      Service #1
    </h2>
    <p>
      Lorem ipsum dolor sit amet, consectetur adipiscing elit. Curabitur quis nunc libero. Nullam tincidunt lectus ut maximus dignissim. Mauris et risus in metus ornare convallis. Nullam in orci leo. Aenean sit amet arcu vitae ipsum mollis blandit. Pellentesque pellentesque sapien mollis, pharetra magna sit amet, varius eros. Donec eu dui ullamcorper ligula rutrum tincidunt. Sed eu venenatis mi. Nullam rutrum sagittis tortor, id feugiat dolor commodo a. Nullam venenatis rutrum elementum. Vestibulum vehicula lectus in nisl porta scelerisque at id metus. Sed porttitor purus eu cursus porta. Donec suscipit semper ligula, ut eleifend nunc commodo ut. Aenean sed mauris molestie, gravida nisi et, lobortis sapien. Nullam et massa mi. 
    </p>
     
    <p>
      Nulla a sem sit amet nisl interdum tincidunt. Fusce pellentesque nec risus et lacinia. Donec dignissim enim ac vehicula malesuada. Lorem ipsum dolor sit amet, consectetur adipiscing elit. Etiam at varius augue. Phasellus varius, sem ac posuere mattis, augue turpis sollicitudin arcu, ut tincidunt arcu tellus vel sapien. Nulla sagittis auctor velit, ut tristique nulla sollicitudin et. Duis ac purus efficitur, congue purus egestas, ultricies quam. Nulla tincidunt nibh risus, vel semper lacus consectetur quis. Aenean luctus enim in justo rhoncus dignissim. Donec pellentesque fermentum sapien vitae aliquam.
    </p>
    <h2>
      Service #1
    </h2>
    <p>
      Lorem ipsum dolor sit amet, consectetur adipiscing elit. Curabitur quis nunc libero. Nullam tincidunt lectus ut maximus dignissim. Mauris et risus in metus ornare convallis. Nullam in orci leo. Aenean sit amet arcu vitae ipsum mollis blandit. Pellentesque pellentesque sapien mollis, pharetra magna sit amet, varius eros. Donec eu dui ullamcorper ligula rutrum tincidunt. Sed eu venenatis mi. Nullam rutrum sagittis tortor, id feugiat dolor commodo a. Nullam venenatis rutrum elementum. Vestibulum vehicula lectus in nisl porta scelerisque at id metus. Sed porttitor purus eu cursus porta. Donec suscipit semper ligula, ut eleifend nunc commodo ut. Aenean sed mauris molestie, gravida nisi et, lobortis sapien. Nullam et massa mi. 
    </p>
     
    <p>
      Nulla a sem sit amet nisl interdum tincidunt. Fusce pellentesque nec risus et lacinia. Donec dignissim enim ac vehicula malesuada. Lorem ipsum dolor sit amet, consectetur adipiscing elit. Etiam at varius augue. Phasellus varius, sem ac posuere mattis, augue turpis sollicitudin arcu, ut tincidunt arcu tellus vel sapien. Nulla sagittis auctor velit, ut tristique nulla sollicitudin et. Duis ac purus efficitur, congue purus egestas, ultricies quam. Nulla tincidunt nibh risus, vel semper lacus consectetur quis. Aenean luctus enim in justo rhoncus dignissim. Donec pellentesque fermentum sapien vitae aliquam.
    </p>
  </div>
</div>
`;
